### PR TITLE
ISSUE-1006 Fix CalendarEvent reply extension throw serverFail error

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/1006

### CODE REVIEW

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by quantranhong1999** [comment]
Can you deploy this fix on preprod? @tddang-linagora may appreciate it :-)

**Comment by Arsnael** [comment]
https://james-jenkins.lin-saas.com/job/Tmail%20build/job/PR-1011/1/

Related

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala [hunk 1]
```diff
@@ -243,6 +246,10 @@ class I18NCalendarEventReplyMessageGenerator(fileSystem: FileSystem, i18nEmlDire
 
   import I18NCalendarEventReplyMessageGenerator._
 
+  private val initCheck: Unit = {
+    Preconditions.checkNotNull(i18nEmlDirectory, s"JMAP CalendarEvent needs a `${I18N_MAIL_TEMPLATE_LOCATION_PROPERTY}` entry".asInstanceOf[Object])
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala:250]: I didnt notice this at first but why do we force having this parameter defined here? I thought we agreed initially that if it's missing, a default location with default templates would be used?

That threw me a guice injection issue when trying to deploy it on staging

Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala:250]: > The throwing error when Jmap server startup will help us avoid that

I thought we set a default template in the default configuration of TMail. Or we miss that?

Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala:250]: Just checked the jib configuration
We already have the default "/root/eml-template" for that 
let me update the code

Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala]: updated & forced push & changed commit message

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala [hunk 2]
```diff
@@ -54,10 +56,10 @@ class CalendarEventReplyPerformer @Inject()(blobCalendarResolver: BlobCalendarRe
                                             supportedLanguage: CalendarEventReplySupportedLanguage,
                                             usersRepository: UsersRepository) extends Startable {
 
-  private val mailReplyGenerator: CalendarEventMailReplyGenerator = Try(jmapConfiguration.getString(I18N_MAIL_TEMPLATE_LOCATION_PROPERTY))
+  private val mailReplyGenerator: CalendarEventMailReplyGenerator = Try(jmapConfiguration.getString(I18N_MAIL_TEMPLATE_LOCATION_PROPERTY, I18N_MAIL_TEMPLATE_LOCATION_DEFAULT))
     .map(i18nEmlDirectory => new I18NCalendarEventReplyMessageGenerator(fileSystem, i18nEmlDirectory)) match {
     case Success(value) => new CalendarEventMailReplyGenerator(value)
-    case Failure(_) => throw new MissingArgumentException("JMAP CalendarEvent needs a " + I18N_MAIL_TEMPLATE_LOCATION_PROPERTY + " entry")
+    case Failure(error) => throw error;
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala:62]: Trailing `;` should not be needed in scala, right?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala [hunk 3]
```diff
@@ -99,7 +101,9 @@ class CalendarEventReplyPerformer @Inject()(blobCalendarResolver: BlobCalendarRe
           .subscribeOn(Schedulers.boundedElastic()))
         .`then`(SMono.just(CalendarEventReplyResults(done = BlobIds(Seq(blobId.value))))))
       .onErrorResume({
-        case _: InvalidCalendarFileException => SMono.just(CalendarEventReplyResults.notDone(blobId))
+        case e@(_: InvalidCalendarFileException | _: IllegalArgumentException) =>
+          LOGGER.error("Error when generate reply mail for {}: {}", mailboxSession.getUser.asString(), e.getMessage)
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala:105]: `error` log for badly formatted user data seems excessive, correct?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/CalendarEventReplyGenerator.scala [hunk 4]
```diff
@@ -49,6 +45,18 @@ object CalendarEventReplyGenerator {
         propertyList
       }).get
 
+    val validationResult: Seq[ValidationResult] = replyProperty.asScala
+      .map(_.validate())
+      .filter(_.hasErrors)
+      .toSeq
+
+    Preconditions.checkArgument(validationResult.isEmpty, s"Invalidate calendar event: ${validationResult.flatMap(_.getEntries.asScala.toSeq).map(error => error.getContext + ": " + error.getMessage).mkString(";")}".asInstanceOf[Object])
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/CalendarEventReplyGenerator.scala:53]: Use string interpolation

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala [hunk 5]
```diff
@@ -54,10 +56,10 @@ class CalendarEventReplyPerformer @Inject()(blobCalendarResolver: BlobCalendarRe
                                             supportedLanguage: CalendarEventReplySupportedLanguage,
                                             usersRepository: UsersRepository) extends Startable {
 
-  private val mailReplyGenerator: CalendarEventMailReplyGenerator = Try(jmapConfiguration.getString(I18N_MAIL_TEMPLATE_LOCATION_PROPERTY))
+  private val mailReplyGenerator: CalendarEventMailReplyGenerator = Try(jmapConfiguration.getString(I18N_MAIL_TEMPLATE_LOCATION_PROPERTY, I18N_MAIL_TEMPLATE_LOCATION_DEFAULT))
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CalendarEventReplyPerformer.scala:59]: <3

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
How about over compatibility modes?

Especially the outlook one can be handy for CNB...

Don't you mind @vttranlina to ensure there is a test for those Outlook events?

**Comment by chibenwa** [comment]
> we can declare in jvm.properties

I tried on tmail.linagora.com but it was not successful.

**Comment by vttranlina** [comment]
> > we can declare in jvm.properties
> 
> I tried on tmail.linagora.com but it was not successful.

I just tested
It works on my local

I added the property: `ical4j.unfolding.relaxed=true` to jvm.properties
And print the `CompatibilityHints.isHintEnabled(CompatibilityHints.KEY_RELAXED_UNFOLDING)` value in `CalendarEventParsed`, it accepts the value from `jvm.properties`

**Comment by Arsnael** [comment]
> I just tested It works on my local

With the code changes of this PR needed, correct?

**Comment by vttranlina** [comment]
> > I just tested It works on my local
> 
> With the code changes of this PR needed, correct?

Not need, 
This PR just set `ical4j.unfolding.relaxed=true`  by default.  (before that, ical4j set `false` by default)

**Pull request merged**